// @angular/fire/esm2015/database/interfaces.js
/**
 * @fileoverview added by tsickle
 * Generated from: interfaces.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 * @template T
 */
export function AngularFireList() { }
if (false) {
    /** @type {?} */
    AngularFireList.prototype.query;
    /**
     * @param {?=} events
     * @param {?=} options
     * @return {?}
     */
    AngularFireList.prototype.valueChanges = function (events, options) { };
    /**
     * @template K
     * @param {?=} events
     * @param {?=} options
     * @return {?}
     */
    AngularFireList.prototype.valueChanges = function (events, options) { };
    /**
     * @param {?=} events
     * @return {?}
     */
    AngularFireList.prototype.snapshotChanges = function (events) { };
    /**
     * @param {?=} events
     * @return {?}
     */
    AngularFireList.prototype.stateChanges = function (events) { };
    /**
     * @param {?=} events
     * @return {?}
     */
    AngularFireList.prototype.auditTrail = function (events) { };
    /**
     * @param {?} item
     * @param {?} data
     * @return {?}
     */
    AngularFireList.prototype.update = function (item, data) { };
    /**
     * @param {?} item
     * @param {?} data
     * @return {?}
     */
    AngularFireList.prototype.set = function (item, data) { };
    /**
     * @param {?} data
     * @return {?}
     */
    AngularFireList.prototype.push = function (data) { };
    /**
     * @param {?=} item
     * @return {?}
     */
    AngularFireList.prototype.remove = function (item) { };
}
/**
 * @record
 * @template T
 */
export function AngularFireObject() { }
if (false) {
    /** @type {?} */
    AngularFireObject.prototype.query;
    /**
     * @return {?}
     */
    AngularFireObject.prototype.valueChanges = function () { };
    /**
     * @return {?}
     */
    AngularFireObject.prototype.snapshotChanges = function () { };
    /**
     * @param {?} data
     * @return {?}
     */
    AngularFireObject.prototype.update = function (data) { };
    /**
     * @param {?} data
     * @return {?}
     */
    AngularFireObject.prototype.set = function (data) { };
    /**
     * @return {?}
     */
    AngularFireObject.prototype.remove = function () { };
}
/**
 * @record
 */
export function FirebaseOperationCases() { }
if (false) {
    /** @type {?} */
    FirebaseOperationCases.prototype.stringCase;
    /** @type {?|undefined} */
    FirebaseOperationCases.prototype.firebaseCase;
    /** @type {?|undefined} */
    FirebaseOperationCases.prototype.snapshotCase;
    /** @type {?|undefined} */
    FirebaseOperationCases.prototype.unwrappedSnapshotCase;
}
/**
 * @record
 * @template T
 */
export function Action() { }
if (false) {
    /** @type {?} */
    Action.prototype.type;
    /** @type {?} */
    Action.prototype.payload;
}
/**
 * @record
 * @template T
 */
export function AngularFireAction() { }
if (false) {
    /** @type {?} */
    AngularFireAction.prototype.prevKey;
    /** @type {?} */
    AngularFireAction.prototype.key;
}
/**
 * @record
 * @template T
 */
export function DatabaseSnapshotExists() { }
if (false) {
    /**
     * @return {?}
     */
    DatabaseSnapshotExists.prototype.exists = function () { };
    /**
     * @return {?}
     */
    DatabaseSnapshotExists.prototype.val = function () { };
    /**
     * @param {?} action
     * @return {?}
     */
    DatabaseSnapshotExists.prototype.forEach = function (action) { };
}
/**
 * @record
 * @template T
 */
export function DatabaseSnapshotDoesNotExist() { }
if (false) {
    /**
     * @return {?}
     */
    DatabaseSnapshotDoesNotExist.prototype.exists = function () { };
    /**
     * @return {?}
     */
    DatabaseSnapshotDoesNotExist.prototype.val = function () { };
    /**
     * @param {?} action
     * @return {?}
     */
    DatabaseSnapshotDoesNotExist.prototype.forEach = function (action) { };
}
